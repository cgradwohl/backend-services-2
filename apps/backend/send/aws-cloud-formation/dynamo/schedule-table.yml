Resources:
  ScheduleTable:
    Type: AWS::DynamoDB::Table
    Properties:
      AttributeDefinitions:
        - AttributeName: pk
          AttributeType: S
        - AttributeName: sk
          AttributeType: S
        - AttributeName: gsi1pk
          AttributeType: S
        - AttributeName: gsi1sk
          AttributeType: S
      KeySchema:
        - AttributeName: pk
          KeyType: HASH
        - AttributeName: sk
          KeyType: RANGE
      GlobalSecondaryIndexes:
        - IndexName: gsi1
          KeySchema:
          - AttributeName: gsi1pk
            KeyType: HASH
          - AttributeName: gsi1sk
            KeyType: RANGE
          Projection:
            ProjectionType: ALL
      BillingMode: PAY_PER_REQUEST
      PointInTimeRecoverySpecification:
        PointInTimeRecoveryEnabled: true
      TimeToLiveSpecification:
        AttributeName: ttl
        Enabled: true
      StreamSpecification:
        StreamViewType: NEW_AND_OLD_IMAGES # Both the new and the old item images of the item are written to the stream
      Tags:
      - Key: VantaOwner
        Value: ${self:custom.vanta.owner}
      - Key: VantaNonProd
        Value: ${self:custom.vanta.nonProd}
      - Key: VantaDescription
        Value: "Stores send application sequence delays and request scheduling."
      - Key: VantaContainsUserData
        Value: true
      - Key: tags.courier.com/service
        Value: pipeline
