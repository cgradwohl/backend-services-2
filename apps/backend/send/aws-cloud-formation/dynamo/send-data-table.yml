Resources:
  SendDataTable:
    Type: AWS::DynamoDB::Table
    Properties:
      AttributeDefinitions:
        - AttributeName: pk
          AttributeType: S
        - AttributeName: sk
          AttributeType: S
        - AttributeName: gsi1pk
          AttributeType: S
        - AttributeName: gsi1sk
          AttributeType: S
        - AttributeName: gsi2pk
          AttributeType: S
        - AttributeName: gsi2sk
          AttributeType: S
        - AttributeName: gsi3pk
          AttributeType: S
        - AttributeName: gsi3sk
          AttributeType: S
      KeySchema:
        - AttributeName: pk
          KeyType: HASH
        - AttributeName: sk
          KeyType: RANGE
      GlobalSecondaryIndexes:
        - IndexName: gsi1
          KeySchema:
          - AttributeName: gsi1pk
            KeyType: HASH
          - AttributeName: gsi1sk
            KeyType: RANGE
          Projection:
            ProjectionType: ALL
        - IndexName: gsi2
          KeySchema:
          - AttributeName: gsi2pk
            KeyType: HASH
          - AttributeName: gsi2sk
            KeyType: RANGE
          Projection:
            ProjectionType: ALL
        # gsi3 is reserved for the sharded workspace item collections
        # use gsi3 to fetch many Entity by workspaceId
        - IndexName: gsi3
          KeySchema:
          - AttributeName: gsi3pk
            KeyType: HASH
          - AttributeName: gsi3sk
            KeyType: RANGE
          Projection:
            ProjectionType: ALL
      BillingMode: PAY_PER_REQUEST
      PointInTimeRecoverySpecification:
        PointInTimeRecoveryEnabled: true
      # Required for listening to dynamo events
      # StreamSpecification:
      #   StreamViewType: NEW_IMAGE
      Tags:
      - Key: VantaOwner
        Value: ${self:custom.vanta.owner}
      - Key: VantaNonProd
        Value: ${self:custom.vanta.nonProd}
      - Key: VantaDescription
        Value: "Stores send application data model such as client requests and derived entities."
      - Key: VantaContainsUserData
        Value: true
      - Key: tags.courier.com/service
        Value: pipeline
