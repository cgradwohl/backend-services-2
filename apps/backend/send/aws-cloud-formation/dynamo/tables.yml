Resources:
  ActionStreamSequenceTable:
    Type: AWS::DynamoDB::Table
    Properties:
      AttributeDefinitions:
        - AttributeName: sequenceNumber
          AttributeType: S
        - AttributeName: lambdaFunction
          AttributeType: S
      KeySchema:
        - AttributeName: sequenceNumber
          KeyType: HASH
        - AttributeName: lambdaFunction
          KeyType: RANGE
      BillingMode: PAY_PER_REQUEST
      PointInTimeRecoverySpecification:
        PointInTimeRecoveryEnabled: true
      TimeToLiveSpecification:
        AttributeName: ttl
        Enabled: true
      Tags:
      - Key: VantaOwner
        Value: ${self:custom.vanta.owner}
      - Key: VantaNonProd
        Value: ${self:custom.vanta.nonProd}
      - Key: VantaDescription
        Value: "Table to hold successfully processed sequences from SendTrafficStream"
      - Key: VantaContainsUserData
        Value: false # true if contains user data
      - Key: tags.courier.com/service
        Value: pipeline

  ProviderSendStreamSequenceTable:
    Type: AWS::DynamoDB::Table
    Properties:
      AttributeDefinitions:
        - AttributeName: sequenceNumber
          AttributeType: S
        - AttributeName: lambdaFunction
          AttributeType: S
      KeySchema:
        - AttributeName: sequenceNumber
          KeyType: HASH
        - AttributeName: lambdaFunction
          KeyType: RANGE
      BillingMode: PAY_PER_REQUEST
      PointInTimeRecoverySpecification:
        PointInTimeRecoveryEnabled: true
      TimeToLiveSpecification:
        AttributeName: ttl
        Enabled: true
      Tags:
      - Key: VantaOwner
        Value: ${self:custom.vanta.owner}
      - Key: VantaNonProd
        Value: ${self:custom.vanta.nonProd}
      - Key: VantaDescription
        Value: "Table to hold successfully processed sequences from SendTrafficStream"
      - Key: VantaContainsUserData
        Value: false # true if contains user data
      - Key: tags.courier.com/service
        Value: pipeline

  ProviderRenderStreamSequenceTable:
    Type: AWS::DynamoDB::Table
    Properties:
      AttributeDefinitions:
        - AttributeName: sequenceNumber
          AttributeType: S
        - AttributeName: lambdaFunction
          AttributeType: S
      KeySchema:
        - AttributeName: sequenceNumber
          KeyType: HASH
        - AttributeName: lambdaFunction
          KeyType: RANGE
      BillingMode: PAY_PER_REQUEST
      PointInTimeRecoverySpecification:
        PointInTimeRecoveryEnabled: true
      TimeToLiveSpecification:
        AttributeName: ttl
        Enabled: true
      Tags:
      - Key: VantaOwner
        Value: ${self:custom.vanta.owner}
      - Key: VantaNonProd
        Value: ${self:custom.vanta.nonProd}
      - Key: VantaDescription
        Value: "Table to hold successfully processed sequences from SendTrafficStream"
      - Key: VantaContainsUserData
        Value: false # true if contains user data
      - Key: tags.courier.com/service
        Value: pipeline

  RetryMessageSequenceTable:
    Type: AWS::DynamoDB::Table
    Properties:
      AttributeDefinitions:
        - AttributeName: sequenceNumber
          AttributeType: S
        - AttributeName: lambdaFunction
          AttributeType: S
      KeySchema:
        - AttributeName: sequenceNumber
          KeyType: HASH
        - AttributeName: lambdaFunction
          KeyType: RANGE
      BillingMode: PAY_PER_REQUEST
      PointInTimeRecoverySpecification:
        PointInTimeRecoveryEnabled: true
      TimeToLiveSpecification:
        AttributeName: ttl
        Enabled: true
      Tags:
      - Key: VantaOwner
        Value: ${self:custom.vanta.owner}
      - Key: VantaNonProd
        Value: ${self:custom.vanta.nonProd}
      - Key: VantaDescription
        Value: "Table to hold successfully processed sequences from SendTrafficStream"
      - Key: VantaContainsUserData
        Value: false # true if contains user data
      - Key: tags.courier.com/service
        Value: pipeline

  RetryMessageTable:
    Type: AWS::DynamoDB::Table
    Properties:
      AttributeDefinitions:
        # <tenantId>#<requestId | messageId>#<command | actionType>
        # If message needs to be retried during prepare
        # ce79b74a-caad-451c-9bd3-0fd95a2d4ea8#1-617b277e-6b17e47279d2a85a380a61bc#prepare
        - AttributeName: pk
          AttributeType: S
      KeySchema:
        - AttributeName: pk
          KeyType: HASH
      BillingMode: PAY_PER_REQUEST
      PointInTimeRecoverySpecification:
        PointInTimeRecoveryEnabled: true
      TimeToLiveSpecification:
        AttributeName: ttl
        Enabled: true
      StreamSpecification:
        StreamViewType: OLD_IMAGE
      Tags:
      - Key: VantaOwner
        Value: ${self:custom.vanta.owner}
      - Key: VantaNonProd
        Value: ${self:custom.vanta.nonProd}
      - Key: VantaDescription
        Value: "Table that contains data necessary to retry failed messages from send stream"
      - Key: VantaContainsUserData
        Value: false
      - Key: tags.courier.com/service
        Value: pipeline
