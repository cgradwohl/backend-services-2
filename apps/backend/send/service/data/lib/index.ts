export const PARTITION_SHARD_RANGE = 10;
