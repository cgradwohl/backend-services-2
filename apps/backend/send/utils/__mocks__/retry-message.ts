export const retryMessage = jest.fn();
