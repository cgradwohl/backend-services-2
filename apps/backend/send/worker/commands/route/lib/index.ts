export { ChannelHandleFailedError } from "~/lib/send-routing";
export * from "./get-user-routing-preferences";
export * from "./dispatch-route-summary-to-render";
