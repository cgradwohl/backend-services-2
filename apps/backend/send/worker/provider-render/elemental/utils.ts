import {
  Content,
  ElementalContent,
  ElementalContentSugar,
  ElementalNode,
} from "~/api/send/types";

export function getElementalContent(content: Content): ElementalContent {
  const sugar = transformContentToElementalContent(
    content as ElementalContentSugar
  );

  return {
    version: "2022-01-01",
    elements: [
      // transform ElementalContentSugar ("title" and "body") into ElementalContent, and give them precedence, if they exist
      ...(sugar?.elements ?? []),
      ...((content as ElementalContent)?.elements ?? []),
    ],
  };
}

export function isElementalContent(
  content: Content
): content is ElementalContent {
  return "elements" in content;
}

export function transformContentToElementalContent(
  content: ElementalContentSugar
): ElementalContent {
  const elements: ElementalNode[] = Object.keys(content).reduce(
    (result, key) => {
      if (key === "title") {
        return [
          ...result,
          {
            type: "meta",
            title: content.title,
          },
        ];
      }

      if (key === "body") {
        return [
          ...result,
          {
            type: "text",
            content: content.body,
          },
        ];
      }

      return result;
    },
    []
  );

  return {
    version: "2022-01-01",
    elements,
  };
}

export function getTitle(elements: ElementalNode[]): string {
  for (const element of elements) {
    if (element.type === "meta" && "title" in element) {
      return element.title;
    }
  }
  return "";
}
