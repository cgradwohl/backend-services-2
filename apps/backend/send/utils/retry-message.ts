import { nanoid } from "nanoid";
import { put } from "~/lib/dynamo";
import {
  createRetryingEvent,
  createTimedoutEvent,
} from "~/lib/dynamo/event-logs";
import { enqueueByQueueUrl } from "~/lib/enqueue";
import getTtl from "~/lib/get-ttl";
import { CourierLogger } from "~/lib/logger";
import {
  getRouteNode,
  isRouteLeafTimedOut,
  RouteNode,
  RouteNodeAddress,
  RouteTimeoutTable,
  SendTimes,
} from "~/lib/send-routing";
import { failover, FailoverOpts } from "./failover";

const RETRY_DELAYS_BY_RETRY_COUNT = {
  1: 10,
  2: 15,
  3: 20,
  4: 25,
  5: 30,
  6: 60,
  7: 120,
  8: 240,
  9: 480,
  10: 600,
};

const MAX_RETRIES = 25;

interface IRetryMessage {
  streamName: string;
  retryCount?: number;
  [key: string]: any;
}

export interface RetryMessageFailoverOpts extends FailoverOpts {
  timeouts?: RouteTimeoutTable;
  tree?: RouteNode;
}

export async function retryMessage(
  message: IRetryMessage,
  opts?: RetryMessageFailoverOpts
) {
  const retryCount = getRetryCount(message);
  const { logger } = new CourierLogger("retryMessage");
  const { messageId, tenantId } = message;
  const delaySeconds = RETRY_DELAYS_BY_RETRY_COUNT[retryCount];

  // Can no longer retry with current provider. Send back to router for failover
  if (hasExceededMaxRetries(retryCount) || isTimedout(opts)) {
    return await Promise.all([
      createTimedoutEvent(tenantId, messageId, {
        channel: message.channel,
        provider: message.provider,
        time: new Date().toISOString(),
      }),
      failover({ ...opts, messageId }),
    ]);
  }

  await createRetryingEvent(tenantId, messageId, { retryCount });

  const newMessage = {
    ...message,
    retryCount,
    streamName: message.streamName,
  };

  // for first 10 retries, we use DelaySeconds from SQS
  // increasing the delay by 30 seconds for each retry up to 10 retries
  if (retryCount <= 10) {
    logger.debug(
      `retryMessage using SQS Visibility Timeout:- ${delaySeconds}, retry count:-${retryCount}`
    );
    await enqueueByQueueUrl<IRetryMessage>(
      process.env.RETRY_SEND_QUEUE_URL,
      delaySeconds
    )(newMessage);
    return;
  }

  const ttl = getTtl(retryCount);

  logger.debug(
    `retryMessage using Dynamo TTL:- ${ttl}, retry count:-${retryCount}`
  );

  try {
    await put({
      Item: {
        ...newMessage,
        pk: nanoid(),
        ttl,
      },
      TableName: process.env.RETRY_TABLE,
    });
    return;
  } catch (error) {
    logger.error("Retry message put failed", newMessage);
    throw error;
  }
}

interface IsTimedoutOpts {
  tree?: RouteNode;
  times?: SendTimes;
  timeouts?: RouteTimeoutTable;
  address?: RouteNodeAddress;
}

const isTimedout = (opts?: IsTimedoutOpts): boolean => {
  if (!opts) return false;
  const { address, tree, timeouts, times } = opts;
  if (!address || !tree || !timeouts || !times) return false;
  const leaf = getRouteNode(address, tree);

  if (!leaf || leaf.type !== "leaf") {
    console.warn(`Leaf not found for address ${address}`);
    return false;
  }

  return isRouteLeafTimedOut({ leaf, times, timeouts });
};

const getRetryCount = (message: IRetryMessage): number => {
  return message?.retryCount ? message.retryCount + 1 : 1;
};

const hasExceededMaxRetries = (retryCount: number): boolean => {
  return retryCount > MAX_RETRIES;
};
