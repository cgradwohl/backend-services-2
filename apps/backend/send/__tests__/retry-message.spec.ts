import { retryMessage } from "~/send/utils/retry-message";
import * as dynamoLib from "~/lib/dynamo";
import {
  createRetryingEvent,
  createTimedoutEvent,
} from "~/lib/dynamo/event-logs";
import { isRouteLeafTimedOut } from "~/lib/send-routing";

jest.mock("~/lib/send-routing");
jest.mock("~/lib/send-routing/lib/get-route-node");
jest.mock("~/send/service/actions");

const putSpy = jest.spyOn(dynamoLib, "put");
const queueMock = jest.fn();
const mockIsRouteLeafTimedOut = isRouteLeafTimedOut as jest.Mock;

const getMockRetryMessage = (mockRetryCount: number) => ({
  messageId: "test-message-id",
  retryCount: mockRetryCount,
  streamName: "test-send-stream",
  tenantId: "test-tenant",
  type: "render",
});

const mockFailoverOpts = {
  address: [],
  tree: {},
  timeouts: {},
  times: {},
};

jest.mock("~/lib/dynamo", () => ({
  put: jest.fn().mockImplementation((params) => {
    const {
      Item: { ttl, pk, ...rest },
    } = params;
    expect(pk).toBeDefined();
    expect(ttl).toBeGreaterThan(0);
    expect(rest).toMatchSnapshot("should match DynamoDb payload");
  }),
}));

jest.mock("~/lib/dynamo/event-logs", () => ({
  createRetryingEvent: jest.fn(),
  createTimedoutEvent: jest.fn(),
}));

jest.mock("~/lib/enqueue", () => {
  return {
    enqueueByQueueUrl: () => queueMock,
  };
});

describe("retry-message", () => {
  beforeAll(() => {
    process.env.RETRY_SEND_QUEUE_URL = "retry-queue-url";
    process.env.RETRY_TABLE = "retry-table";
    process.env.DESTINATION_STREAM_FOR_RETRY = "test-send-stream";
  });

  beforeEach(jest.clearAllMocks);

  describe("retry failures in send kinesis stream workers", () => {
    it(`should retry using SQS VisibilityTimeout for first 10 retries`, async () => {
      expect.assertions(3);
      await retryMessage(getMockRetryMessage(1));
      expect(queueMock).toHaveBeenCalledTimes(1);
      expect(queueMock).toHaveBeenCalledWith(getMockRetryMessage(2));
      expect(createRetryingEvent).toHaveBeenCalledTimes(1);
    });

    it(`should retry using Dynamo TTL for last 15 retries`, async () => {
      expect.assertions(5); // The dynamo mock has assertions. This caused me much pain and confusion -_-
      await retryMessage(getMockRetryMessage(12));
      expect(putSpy).toHaveBeenCalledTimes(1);
      expect(createRetryingEvent).toHaveBeenCalledTimes(1);
    });
  });

  describe("unretryable", () => {
    it(`should not retry using Dynamo TTL for last 15 retries if timedout`, async () => {
      expect.assertions(4);
      mockIsRouteLeafTimedOut.mockReturnValue(true);
      await retryMessage(getMockRetryMessage(12), mockFailoverOpts as any);
      expect(putSpy).toHaveBeenCalledTimes(0);
      expect(createRetryingEvent).toHaveBeenCalledTimes(0);
      expect(createTimedoutEvent).toHaveBeenCalledTimes(1);
      expect(mockIsRouteLeafTimedOut).toHaveBeenCalledTimes(1);
    });

    it(`should not retry using Dynamo TTL for after 25 retries`, async () => {
      expect.assertions(1);
      await retryMessage(getMockRetryMessage(25));
      expect(createTimedoutEvent).toHaveBeenCalledTimes(1);
    });
  });
});
