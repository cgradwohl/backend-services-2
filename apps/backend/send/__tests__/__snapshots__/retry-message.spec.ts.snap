// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`retry-message retry failures in send kinesis stream workers should retry using Dynamo TTL for last 15 retries: should match DynamoDb payload 1`] = `
Object {
  "messageId": "test-message-id",
  "retryCount": 13,
  "streamName": "test-send-stream",
  "tenantId": "test-tenant",
  "type": "render",
}
`;
